File: doc-examples/hibernate-reactive-example-kotlin/src/main/kotlin/example/BookRepository.kt

package example

import io.micronaut.data.annotation.Id
import io.micronaut.data.annotation.Query
import io.micronaut.data.annotation.Repository
import io.micronaut.data.model.Page
import io.micronaut.data.model.Pageable
import io.micronaut.data.model.Slice
import io.micronaut.data.repository.kotlin.CoroutineCrudRepository
import java.util.function.Consumer
import javax.transaction.Transactional

// tag::repository[]
@Repository // <1>
interface BookRepository : CoroutineCrudRepository<Book, Long> { // <2>

    // tag::read[]
    suspend fun find(title: String): Book

    suspend fun findOne(title: String): BookDTO

    suspend fun findByPagesGreaterThan(pageCount: Int, pageable: Pageable): List<Book>

    suspend fun findByTitleLike(title: String, pageable: Pageable): Page<Book>

    suspend fun list(pageable: Pageable): Slice<Book>
    // end::read[]

    // tag::save[]
    suspend fun save(entity: Book): Book
    // end::save[]

    // tag::inserts[]
    @Query("INSERT INTO Book(title, pages) VALUES (:title, :pages)")
    suspend fun insert(title: String, pages: Int)

    // tag::update[]
    @Transactional
    suspend fun findByIdAndUpdate(id: Long, bookConsumer: Consumer<Book?>) {
        bookConsumer.accept(findById(id))
    }

    suspend fun update(newBook: Book): Book

    suspend fun update(@Id id: Long?, pages: Int)
    // end::update[]

    // tag::delete[]
    suspend fun delete(title: String)
    // end::delete[]
}
// end::repository[]

File: doc-examples/azure-cosmos-example-python/src/main/python/example/BookRepository.py
from typing import Annotated

from micronaut.data.annotation import Id
from micronaut.data.cosmos.annotation import CosmosRepository
from micronaut.data.model import Pageable, Slice
from micronaut.data.repository import CrudRepository

from example.Book import Book
from example.BookDTO import BookDTO


# tag::repository[]
@CosmosRepository  # <1>
class BookRepository(CrudRepository[Book, str]):  # <2>
    # end::repository[]

    # tag::simple[]
    def findByTitle(self, title: str) -> Book: ...

    def getByTitle(self, title: str) -> Book: ...

    def retrieveByTitle(self, title: str) -> Book: ...
    # end::simple[]

    # tag::greaterthan[]
    def findByPagesGreaterThan(self, pageCount: int) -> list[Book]: ...
    # end::greaterthan[]

    # tag::simple-alt[]
    # tag::repository[]
    def find(self, title: str) -> Book: ...
    # end::simple-alt[]
    # end::repository[]

    # tag::pageable[]
    def findAllByPagesGreaterThan(self, pageCount: int, pageable: Pageable) -> list[Book]: ...

    def list(self, pageable: Pageable) -> Slice[Book]: ...
    # end::pageable[]

    # tag::simple-projection[]
    def findTitleByPagesGreaterThan(self, pageCount: int) -> list[str]: ...
    # end::simple-projection[]

    # tag::ordering[]
    def listOrderByTitle(self) -> list[Book]: ...

    def listOrderByTitleDesc(self) -> list[Book]: ...
    # end::ordering[]

    # tag::save[]
    def persist(self, entity: Book) -> Book: ...
    # end::save[]

    # tag::save2[]
    def store(self, title: str, pages: int) -> Book: ...
    # end::save2[]

    # tag::update[]
    def update(self, id: Annotated[str, Id], title: str) -> None: ...
    # end::update[]

    # tag::update2[]
    def updateByTitle(self, title: str, pages: int) -> None: ...
    # end::update2[]

    # tag::deleteall[]
    def deleteAll(self) -> None: ...
    # end::deleteall[]

    # tag::deleteone[]
    def delete(self, title: str) -> None: ...
    # end::deleteone[]

    # tag::dto[]
    def findOne(self, title: str) -> BookDTO: ...
    # end::dto[]
